source/common/common/inline_map_registry.h

```diff
@@ -0,0 +1,423 @@
+      // Initialize the inline entries to nullptr.
+      memset(inline_entries_, 0, registry_.registrationNum() * sizeof(TPtr));
+    }
+
+    template <class ArgK> T* lookupImpl(const ArgK& key) const {
+      // Because the normal string view key is used here, try the normal map entry first.
+      if (auto it = dynamic_entries_.find(key); it != dynamic_entries_.end()) {
+        return it->second.get();
+      }
+
+      if (auto entry_id = inlineLookup(key); entry_id.has_value()) {
+        return inline_entries_[*entry_id];
+      }
+
+      return nullptr;
+    }
+
+    T* lookupImpl(InlineKey key) const {
+#ifndef NDEBUG
+      ASSERT(key.registryId() == registry_.registryId(), "Invalid inline key");
+#endif
+      return inline_entries_[key.inlineId()];
+    }
+
+    void resetInlineMapEntry(uint64_t inline_entry_id, TPtr new_entry = nullptr) {
+      ASSERT(inline_entry_id < registry_.registrationNum());
+      if (auto entry = inline_entries_[inline_entry_id]; entry != nullptr) {
+        // Remove and delete the old valid entry.
+        ASSERT(inline_entries_size_ > 0);
+        --inline_entries_size_;
+        delete entry;
+      }
+
+      if (new_entry != nullptr) {
+        // Append the new valid entry.
+        ++inline_entries_size_;
+      }
+
+      inline_entries_[inline_entry_id] = new_entry.release();
+    }
+
+    template <class ArgK> absl::optional<uint64_t> inlineLookup(const ArgK& key) const {
+      const auto& map_ref = registry_.registrationMap();
+      if (auto iter = map_ref.find(key); iter != map_ref.end()) {
+        return iter->second;
+      }
+      return absl::nullopt;
+    }
+
+    // This is the reference of the registry that the inline map created by. This is used to
+    // validate the inline handle validity and get the inline key set.
+    const InlineMapRegistry& registry_;
+
+    // This is the underlay hash map for the normal entries.
+    UnderlayHashMap dynamic_entries_;
+
+    uint64_t inline_entries_size_{};
+    // This should be the last member of the class and no member should be added after this.
+    T* inline_entries_[];
+  };
+
+  template <class T> using InlineMapPtr = std::unique_ptr<InlineMap<T>>;
+
+  InlineMapRegistry() = default;
+
+  /**
+   * Register a custom inline key. May only be called before finalized().
+   */
+  template <class ArgK> InlineKey registerInlineKey(const ArgK& key) {
+    RELEASE_ASSERT(!finalized_, "Cannot register inline key after finalized()");
+
+    uint64_t inline_id = 0;
+
+    if (auto it = registration_map_.find(key); it != registration_map_.end()) {
+      // If the key is already registered, return the existing key.
+      inline_id = it->second;
+    } else {
+      // If the key is not registered yet, then create a new inline key.
+      inline_id = registration_map_.size();
+      registration_set_.push_back(Key(key));
+      registration_map_[key] = inline_id;
+    }
+
+#ifndef NDEBUG
+    return {inline_id, registry_id_};
```

Comment:
Thanks for ifdefing this out; it helps to separate what you need to build the map vs what you want for debug for review.

Now...can you describe in what scenario you'd need this registry ID and the debug power it gives you?

Reply:
In the debug mode, the registry id will be used to ensure all inline key are used in correct inline map and no inline key is mis-used.

It's possible a key k1 from registry A will have same index with a key k2 from registry B. If k1 is used in the inline map from registry B, it may not crash in the release mode. But it still a risk and potential error.